File: Structural/Flyweight/python/test_flyweight.py
import unittest
from flyweight import FlyweightFactory


class TestFlyweightPattern(unittest.TestCase):
	"""
	Unit tests for the Flyweight pattern implementation.
	"""

	def test_flyweight_reuse(self):
		"""
		Test that the FlyweightFactory reuses existing Flyweights for the same intrinsic state.
		"""
		factory = FlyweightFactory()

		# Retrieve Flyweights with the same intrinsic state
		flyweight1 = factory.get_flyweight("SharedState")
		flyweight2 = factory.get_flyweight("SharedState")

		# Verify that the same Flyweight instance is reused
		self.assertIs(flyweight1, flyweight2)

	def test_flyweight_operation(self):
		"""
		Test that Flyweight operations correctly use intrinsic and extrinsic states.
		"""
		factory = FlyweightFactory()
		flyweight = factory.get_flyweight("State")

		# Capture logs for Flyweight operation
		with self.assertLogs("flyweight", level="INFO") as captured:
			flyweight.operation("ExtrinsicState")

		self.assertIn(
			"INFO:flyweight:ConcreteFlyweight: Intrinsic state = State, Extrinsic state = ExtrinsicState",
			captured.output,
		)

	def test_flyweight_list(self):
		"""
		Test that the FlyweightFactory correctly lists all Flyweights.
		"""
		factory = FlyweightFactory()

		# Add Flyweights
		factory.get_flyweight("State1")
		factory.get_flyweight("State2")
		factory.get_flyweight("State1")  # Reuses existing Flyweight

		# Capture logs for listing Flyweights
		with self.assertLogs("flyweight", level="INFO") as captured:
			factory.list_flyweights()

		self.assertIn("INFO:flyweight:Flyweights in Factory:", captured.output)
		self.assertIn("INFO:flyweight: - State1", captured.output)
		self.assertIn("INFO:flyweight: - State2", captured.output)


if __name__ == "__main__":
	unittest.main()

File: Structural/Flyweight/cpp/test_flyweight.cpp
#include "flyweight.hpp"
#include <gtest/gtest.h>

/**
 * @brief Test the creation and reuse of Flyweight objects in the FlyweightFactory.
 */
TEST(FlyweightPatternTest, FlyweightCreationAndReuse) {
	FlyweightFactory factory;

	// Retrieve Flyweights with different intrinsic states
	auto fw1 = factory.getFlyweight("State1");
	auto fw2 = factory.getFlyweight("State2");

	// Reuse the Flyweight with intrinsic state "State1"
	auto fw3 = factory.getFlyweight("State1");

	// Assert that the Flyweight instances with the same intrinsic state are reused
	EXPECT_EQ(fw1, fw3);
	EXPECT_NE(fw1, fw2);

	// List all Flyweights in the factory
	factory.listFlyweights();
}

/**
 * @brief Test Flyweight operation with both intrinsic and extrinsic states.
 */
TEST(FlyweightPatternTest, FlyweightOperation) {
	FlyweightFactory factory;

	// Create Flyweights
	auto fw1 = factory.getFlyweight("SharedState");

	// Perform operations with extrinsic states
	testing::internal::CaptureStdout();
	fw1->operation("ExtrinsicState1");
	fw1->operation("ExtrinsicState2");
	std::string output = testing::internal::GetCapturedStdout();

	// Verify the output
	EXPECT_NE(output.find("ConcreteFlyweight with intrinsic state: SharedState and extrinsic state: ExtrinsicState1"), std::string::npos);
	EXPECT_NE(output.find("ConcreteFlyweight with intrinsic state: SharedState and extrinsic state: ExtrinsicState2"), std::string::npos);
}

/**
 * @brief Test listing of Flyweights in the factory.
 */
TEST(FlyweightPatternTest, ListFlyweights) {
	FlyweightFactory factory;

	// Add multiple Flyweights
	factory.getFlyweight("State1");
	factory.getFlyweight("State2");

	// Capture the output of listing Flyweights
	testing::internal::CaptureStdout();
	factory.listFlyweights();
	std::string output = testing::internal::GetCapturedStdout();

	// Verify the output contains the expected states
	EXPECT_NE(output.find("State1"), std::string::npos);
	EXPECT_NE(output.find("State2"), std::string::npos);
}
